scripts/lint-style.lean

```diff
@@ -12,10 +13,58 @@ import Cli.Basic
 
 This files defines the `lint-style` executable which runs all text-based style linters.
 The linters themselves are defined in `Mathlib.Tactic.Linter.TextBased`.
+
+In addition, this checks that `Mathlib.Init` is (transitively) imported in all of mathlib.
 -/
 
 open Cli Mathlib.Linter.TextBased
 
+/-- Check that `Mathlib.Init` is transitively imported in all of Mathlib. -/
+def checkInitImports : IO Bool := do
+  -- Find any file in the Mathlib directory which does not contain any Mathlib import.
+  -- The current check (for any line starting with "import Mathlib") is not perfect
+  -- (for instance, it would detect a multi-line comment containing the line import Mathlib),
+  -- but is good enough in practice.
+  let mut modulesWithoutMathlibImports := #[]
+  -- We simply parse `Mathlib.lean`; this removes the need for external tools like grep.
+  let allModules := (← IO.FS.lines "Mathlib.lean").map (·.stripPrefix "import ")
+    |>.filter (·.startsWith "Mathlib")
+  for module in allModules do
+    let path := (System.mkFilePath (module.split (· == '.'))).addExtension "lean"
+    let hasMathlibImport := (← IO.FS.lines path).any fun s ↦ s.startsWith "import Mathlib"
+    if !hasMathlibImport then
+      modulesWithoutMathlibImports := modulesWithoutMathlibImports.push module
+
+  -- We check that each of these is imported in `Mathlib/Init.lean`.
+  let initImports := (← IO.FS.lines "Mathlib/Init.lean").filter (·.startsWith "import ")
```

Comment:
If the linter is ran on a computer with path-separator different from `/`. then the path `Mathlib/Init.lean` would not be recognized, right?

I also wonder about the `lean` extension in `allModules`.

Reply:
Good catch! Should somebody write a linter against scripts doing manual path manipulation like this? At least for mathlib scripts, gradually forcing less use of duct tape could be useful :-)

`allModules` parses `Mathlib.lean` --- lines in it look like `import Mathlib.Algebra.Algebra.Pi`, so `allModules` contains entries like `Mathlib.Algebra.Algebra.Pi`, i.e. is a list of module names. Does that answer your question?

Reply:
I was more wondering about `Mathlib.lean` and whether there might be some system-dependent issue in using the extension "manually" instead of going via some `addExtension` API.  It looks fine as is, though.